src/main/java/com/google/roomies/servlets/ListingsServlet.java

```diff
@@ -80,4 +84,60 @@ public void doPost(HttpServletRequest request, HttpServletResponse response) {
         response.setStatus(HttpServletResponse.SC_BAD_REQUEST);
     }
   }
+
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response)
+      throws IOException {
+    database = DatabaseFactory.getDatabase();
+    try {
+      List<Listing> listings = getAllListingsFromCollection();
+
+      response.setContentType("application/json");
+      response.getWriter().println(convertToJsonUsingGson(listings));
+    } catch (Exception e) {
+      System.err.println("Error fetching listings: " + e);
+      response.setStatus(HttpServletResponse.SC_BAD_REQUEST);
+      response.setContentType("text/html");
+      response.getWriter().println("request failed");
+    }
+  }
+
+  private String convertToJsonUsingGson(List data) {
+    Gson gson = new GsonBuilder()
+                  .registerTypeAdapter(Money.class, new MoneySerializer())
+                  .create();
+    return gson.toJson(data);
+  }
+
+  /**
+  * Gets all listings from the listing collection in Firestore.
+  * 
+  * @return list of Listing instances
+  */
+  private List<Listing> getAllListingsFromCollection() throws Exception {
+    List<QueryDocumentSnapshot> documents = 
+      database.getAllDocumentsInCollection(LISTING_COLLECTION_NAME).get().getDocuments();
+
+    return StreamSupport.stream(documents.spliterator(), /* parallel= */ false)
+      .map(this::getListingFromDocument)
+      .flatMap(Optional::stream)
```

Comment:
Does this work if you make the first map a flatMap directly, and omit the second one? Would work in scala, don't remember if Java will be happy that way.

Reply:
Changed to using one flatMap call.